osu.Game/Online/HubClientConnector.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Client;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using osu.Framework;
+using osu.Framework.Bindables;
+using osu.Framework.Logging;
+using osu.Game.Online.API;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A component that maintains over a hub connection between client and server.
+    /// </summary>
+    public class HubClientConnector : IDisposable
+    {
+        /// <summary>
+        /// Invoked whenever a new hub connection is built.
+        /// </summary>
+        public Action<HubConnection>? OnNewConnection;
+
+        private readonly string clientName;
+        private readonly string endpoint;
+        private readonly IAPIProvider? api;
+
+        /// <summary>
+        /// The current connection opened by this connector.
+        /// </summary>
+        public HubConnection? CurrentConnection { get; private set; }
+
+        /// <summary>
+        /// Whether this is connected to the hub, use <see cref="CurrentConnection"/> to access the connection, if this is <c>true</c>.
+        /// </summary>
+        public IBindable<bool> IsConnected => isConnected;
+
+        private readonly Bindable<bool> isConnected = new Bindable<bool>();
+        private readonly SemaphoreSlim connectionLock = new SemaphoreSlim(1);
+        private CancellationTokenSource connectCancelSource = new CancellationTokenSource();
+
+        private readonly IBindable<APIState> apiState = new Bindable<APIState>();
+
+        /// <summary>
+        /// Constructs a new <see cref="HubClientConnector"/>.
+        /// </summary>
+        /// <param name="clientName">The name of the client this connector connects for, used for logging.</param>
+        /// <param name="endpoint">The endpoint to the hub.</param>
+        /// <param name="api">The API provider for listening to state changes, or null to not listen.</param>
```

Comment:
```suggestion
        /// <param name="api">An API provider used to react to connection state changes.</param>
```

Also what's with the null case? It looks like it will never connect? That's not what the xmldoc reads.

Reply:
right, does this sound good?
```suggestion
        /// <param name="api"> An API provider used to react to connection state changes, or null to not establish connection at all (for testing purposes).</param>
```